### ISSUE TITLE
[Question] How to handle possible duplicate "notification" event for notification with `content-available=1`

### ISSUE THREAD

**Issue by tamakisquare** [phonegap/phonegap-plugin-push#1230]
First of all, this is not an issue but a question on how to use the awesome "phonegap-plugin-push" plugin to meet my requirements. My question is specifically for iOS platform, as I haven't tried the plugin with Android just yet.

For notification with `content-available=1`, the **notification** event handler will potentially get called at two times. First is when the push notification is received while the app is in the background. And then when the user taps on the push notification to open the app, the event handler gets called again. The `data` argument provided to the event handler is the same in both occurrences. My question is how can my code distinguish one case from the other to avoid duplicate processing. The only way that I can think of is relying on Cordova events - _"pause"_ and _"resume"_ - to detach/attach handlers to the **notification** event. I think that is doable but I am looking forward to see if there are other more straightforward solutions. Something supported by the "phonegap-plugin-push" plugin would be nice. Thanks.

**Comment by macdonst** [comment]
@tamakisquare in the push payload you send to your app include a unique identifier. Then in local storage you can indicate that the payload has already been processed. That way you handle it in the background and can safely ignore it when the user taps on the notification. 

Make sense?

**Comment by tamakisquare** [comment]
Sounds good! Thanks for your great work and help, @macdonst . Cheers.

**Comment by macdonst** [comment]
@tamakisquare kk, I'm going to re-open this though as a doc issue as more than one person has asked so I should add it to the docs.

**Comment by tamakisquare** [comment]
@macdonst , since you mentioned about documentation, here are some notes of mine on how the plugin behaves in different scenarios. Since it's written from a user's perspective, I thought they or part of them might be of value to the docs. Let me know if there's anything that I can help with this regard. Cheers.

## Cases Observed with **iOS** using **phonegap-plugin-push**
- When a notification is received, regardless whether or not the user has granted your app the permission for Notifications
    - With your app in the foreground
        - "notification" event handler is called with `data.additionalData.foreground=true` and `data.additionalData.coldStart=false`
    - With your app is in background *AND* `content-available` set to 1
        - "notification" event handler is called with `data.additionalData.foreground=false` and `data.additionalData.coldStart=false`
- User's permission for Notifications affects notifications from your app only in the following ways:
    1.  Whether to display the notifications in Notification Center
    2. Whether to display the notifications on Lock Screen
    3. Whether to alert upon receiving the notifications
    4. Whether to show app icon badge
- When there's some notification for your app in Notification Centre
    - Tap on your app icon (regardless whether or not the app is in the background)
        - no "notification" event
    - Tap on the push notification
        - With your app in the background
            - "notification" event with `data.additionalData.foreground=false` and `data.additionalData.coldStart=false`
        - With your app closed (ie. Not even in the background)
            - "notification" event with `data.additionalData.foreground=false` and `data.additionalData.coldStart=true`
- Notification with `content-available=1` will potentially trigger the *notification* event twice
    - With assigning a unique identifier to each notification, use localStorage to record down the notification that has been processed (1st time) and this will be used to avoid repeating the same processing on the same notification (2nd time).

**Comment by jasonatt** [comment]
@macdonst I am getting the same issue which is very random however I did not understand where I would use localstorage and do the check in the background.

**Comment by tamakisquare** [comment]
@jasonatt - Here you go. Hope it makes things clearer for you.

```javascript
if (data.additionalData.foreground) {
  /**
   * This block is reached when a push notification is received when the app is in foreground.
   * Depending on the design of your app, you might want to save the incoming data to the localstorage here.
   * That's not really what your question is about, but I thought mentioning it would give a better picture.
   */
}
else {
  if (data.additionalData.coldstart) {
    /**
     * This block is reached when a push notification is tapped on *AND*
     * the app is closed (ie. Not even in background).
     */
  }
  else {
    /**
     * The following block is reached in the following two scenarios:
     * 1. A push notification is tapped on *AND* the app is in background.
     * 2. A push notification with "content-available=1" is received while the app is in background. (iOS only)
     *
     * Therefore, for iOS devices, both of the above two scenarios can occur for an individual push notification.
     *
     * For Android, it's always the first scenario as the other one is not applicable.
     */

    // Here we check for the existence of the incoming data in the localstorage and, with that, we determine
    // which scenario of the two possible is happening.
    var pastPushSavedID = window.localStorage.getItem("pastPushSavedID");

    if (pastPushSavedID) {
      // `pastPushSavedID` is found in localStorage and thus this is scenario 1 mentioned above.
    }
    else {
      // `pastPushSavedID` is `null` and thus this is scenario 2 mentioned above.
    }
  }
}
```

**Comment by jasonatt** [comment]
@tamakisquare thank you for the detailed reply mush obliged.

Just to confirm that goes in the  push.on notification correct?

**Comment by tamakisquare** [comment]
@jasonatt Yes, the whole thing should go in the handling function for `push.on('notification')`. I am not sure what you mean by "notification areas". Do you mean the notification center at the os-level for unread push notifications? Or are you referring to the notification area, assuming there is one, within the app you are developing?

**Comment by jasonatt** [comment]
@tamakisquare yes os level. If I am in the foreground with your code I can detect if the push was already received by checking the id and I can take action from there however this is not working in the background as I still get the notification with sound and alert (os level)

**Comment by tamakisquare** [comment]
@jasonatt If I understand you correctly, what you are saying is that you want your app to be able to interfere with how device handles push notification. As far as I know, that's not possible because it's beyond the boundary of what an app can do. Push notification behaviour can only be controlled by its sender, via APNs or FCM or any other push notification services, and the device local settings. 

As for the _"notification"_ event being fired twice when your app is in the foreground, I have never got that issue with iOS 9 nor Android 4.4.3.

**Comment by jasonatt** [comment]
@tamakisquare 

Thanks for the update.

Yes I know it was not possible was just checking if this got changed :) 

To be clear the issue is not from the plugin as after some debugging I noticed that the server was actually sending them double :/

Thanks again!

Thanks
